app/scheduled/circulars/index.ts

```diff
@@ -0,0 +1,66 @@
+/*!
+ * Copyright © 2023 United States Government as represented by the
+ * Administrator of the National Aeronautics and Space Administration.
+ * All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import { tables } from '@architect/functions'
+import type { DynamoDBDocument } from '@aws-sdk/lib-dynamodb'
+import { paginateScan } from '@aws-sdk/lib-dynamodb'
+
+import type { CircularAction, CircularActionContext } from './circularAction'
+import { finalizeTar, setupTar, uploadJsonTar, uploadTxtTar } from './uploadTar'
+import { type Circular } from '~/routes/circulars/circulars.lib'
+
+async function mapCirculars(
+  ...actions: CircularAction<CircularActionContext>[]
+) {
+  for await (const circularArray of getAllRecords()) {
+    for (const action of actions) {
+      // the context is based on either the initial setup context or the context of the callback return
+      const context = action.currentContext
+        ? action.currentContext
+        : action.initialContext()
+      const results = await action.action(circularArray, context)
+      action.currentContext = results
+    }
+  }
+  for (const action of actions) {
+    // the finalize context is based the final results of the callback function
+    await action.finalize(action.currentContext)
+  }
```

Comment:
Use `Promise.all` to do as much work as possible concurrently.

```suggestion
  const contexts = await Promise.all(actions.map({initialize} => initialize()))
  for await (const circulars of getAllRecords()) {
    for (const action of actions) {
      await Promise.all(actions.map(({action}, i) => action(circulars, contexts[i])))
    }
  }
  await Promise.all(actions.map(({finalize}, i) => finalize(contexts[i])))
```